WIP: Switch directv to async and improve discovery

ehendrix23 opened the pull request:
## Description:

WIP PR created to receive feedback and initial review.

Changed how discovery works for client (RVU) devices. These devices are discovered from the main DVR, issue is that they are not reported unless they are turned on. Hence HASS discovery cannot discover them. Now a task is scheduled instead when the main DVR entity is created that will execute a discovery for client devices connected to that main DVR entity.

Changed the platform to async, using async_add_executor_job to call the API for the device (as that is not async). Did this to better supported the discovery by using async in HASS.

Fix where entity sometimes would be marked unavailable if we would check status just when a client device would be turned off (or on).

Further improvement to determine if the device is playing or paused.

test_directv.py received small update to support testing this better discovery. PR #18590 however has everything for the test and this PR will be dependent on #18590 being merged first.

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):** home-assistant/home-assistant.io#<home-assistant.io PR number goes here>

## Example entry for `configuration.yaml` (if applicable):
```yaml
media_player:
  - platform: directv
    host: 192.168.1.2
    port: 8080
    name: Family_Room
    device: 0
    discover_clients: true
    client_discover_interval: 20
  - platform: directv
    host: 192.168.1.2
    port: 8080
    name: Master Bedroom
    device: 0
    discover_clients: false
```

## Checklist:
  - [X] The code change is tested and works locally.
  - [X] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [X] There is no commented out code in this PR.

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code does not interact with devices:
  - [X] Tests have been added to verify that the new code works.

ehendrix23 committed:
Enhancements for DirecTV media player

Following enhancements have been made:

1. Added debug logging
2. Added ability to change channel using select_source service of the remote platform.
3. State will now show paused if a recorded program is paused, for live TV playing will always be returned.
4. Added the following attributes:
    a. media_position: current position of the media (in seconds)
    b. media_position_updated_at: timestamp when media_position was updated.
   c. source: current source (channel).
   d. media_isbeingrecorded: if current media is being recorded or not.
   e. media_rating: TV/Movie rating of the media
   f. media_recorded: if current media is recorded or live TV
   g. media_starttime: Timestamp media was aired

Reordered properties to follow same order as how they are in __init__.py of remote platform.

ehendrix23 committed:
Fixed error and cleaned up few items

Fixed an issue when determining if a program is recorded or not.
Cleaned up some coding.

ehendrix23 committed:
Added available property

Added available property

ehendrix23 committed:
Disable feature TURN_ON and TURN_OFF for DVR clients

Disable the feature turn_on and turn_off for DVR clients.

ehendrix23 committed:
Move discover of clients to update for the main entity

Move the discovering of clients to the update process. Clients only report in when they are powered on. HASS discovery only calls the setup_platform once when the main DVR unit is discovered, This means that any clients that are not powered on will not be added and never will.

ehendrix23 committed:
Fixed  lint issues

Fixed liny issues.

ehendrix23 committed:
Changed to async

Changed to async and then have discovery done through a scheduled task instead.

ehendrix23 committed:
Added entity_id to debug log entries

Added entity_id to debug log entries

ehendrix23 committed:
Small update to a debug log entry

Small update to a debug log entry

ehendrix23 committed:
Fix Travis error

Trying to fix Travis error.

ehendrix23 committed:
Fixed test issue

Fixed test issue

ehendrix23 committed:
Further improvements to available state

Additional improvement to available state.

ehendrix23 committed:
Added some more error handling

Some more error handling in case a request fails.

ehendrix23 committed:
Rebase with dev

Rebase with dev

ehendrix23 committed:
Small update to use constant

Small update to use constant

MartinHjelmare commented on homeassistant/components/media_player/directv.py:
```diff
@@ -47,105 +54,238 @@
+                self.hass.data.setdefault(DATA_DIRECTV, set()).add(
+                    (new_device[CONF_HOST], new_device[CONF_DEVICE]))
+
+            self._async_add_entities(dtvs)
```
We shouldn't add new entities from within another entity. That's really weird. If we need to add entities after platform setup, make another instance that handles this that is not an entity.

MartinHjelmare reviewed (commented):
What's the benefit of using asyncio in this platform when the directv library isn't using asyncio?

MartinHjelmare commented:
Please focus on one problem per PR. Don't try to fix everything in one go. This PR is too big without needing to be that.

ehendrix23 committed:
Further improvements.

-) Re-arranged code a bit to increase readability
-) Made code simpler by using partial function
-) unavailable will now be set for the entity if the 1st time reported was a minute or longer ago.

ehendrix23 commented:
@MartinHjelmare, 

For splitting PR. That was my plan but for some items (i.e. async, discovery) I already wanted to see for some initial overall comments on approach while I am still testing/developing. I'm always letting it run for a few days here allowing me to see the different behaviors as I'm using the equipment. :-) 

For adding entities from within an entity. I had discussed this with Paul briefly as well before. The issue really is that the client devices do not show up to be discoverable unless they are turned on. And discovery of them is by calling an API to the main DVR. However, I do also see what you're saying and will work on getting it split off.

Benefit would be overall improvement for HASS by reducing the threading performance penalty in Python. Similar like what was done for other platforms (i.e. Xiaomo_milo light component).

MartinHjelmare commented:
Using the async home assistant api for sync libraries risk the opposite, creating more thread switching overhead.

I think we should keep it sync until the library is using asyncio.

ehendrix23 committed:
Client discovery in seperate class

Moved client discovery into its own class.

ehendrix23 commented:
HASS itself runs on the main thread with the event loop. When there is a sync task then HASS schedules this on another thread for execution.

For example, schedule_update_task ends up executing run_in_executor to execute the update() method in a separate thread. Thus everything in update() is done within the separate thread.

With switching to async, what is done in the separate thread (by using async_add_executor_job which then runs run_in_executor also) is not the complete update() but instead only the synchronous lib call.
Hence the amount of time in a separate thread is going to be less, number of executed instructions less, and hence less potential thread switching when being executed

Thus I would think that overall the preference order would be:
1. async the complete platform (including lib)
2. async platform, use sync for lib calls as reason shown in "Calling sync functions from async" (https://developers.home-assistant.io/docs/en/asyncio_working_with_async.html)
3. Everything in separate thread 

Similar as I for example see what was done in component Xiaomo_milo.

Trying to understand as it would seem to me that having part as async would be better then none at all, and I can see some components that are using async and then calling the library in a sync manner. 

Might not be a bad idea to provide this information with the HASS documentation from a standard perspective (i.e. no async unless library itself is async, or asking for developers to use async and then call libraries using async_add_executor_job). :-)

ehendrix23 commented:
FYI, pushed a change making the discovery a different class and scheduling it when the platform is setup for the main DVR.

ehendrix23 committed:
Update client discover to use received ID

Using host for preventing duplicates can still result if configuration as actual hostname and not an IP.
Using receiver ID to ensure no duplicate is added, host (hostname or IP) used as fall-back.

ehendrix23 committed:
Update required directy version

Update required directpy version to 0.6
